### DISCUSSION TITLE
[avaram] Skip save on specified field if this fields content is blank in before_save

### DISCUSSION THREAD

**Discussion by zw963** [luckyframework/lucky#1701] [Q&A] [answered]
I use avram with kemal.

Following is a example.

```cr
class UpdateUser < User::SaveOperation
  permit_columns licenses_allowed_creation_times, password
  
  before_save do
    if licenses_allowed_creation_times.value.blank? || !licenses_allowed_creation_times.changed?
	  # Need skip save on licenses_allowed_creation_times column here
    end
	
	# others password update logic here.
  end
end
```

I guess, if no add before_save hook, the value from `licenses_allowed_creation_times` column will always be saved to db, right?

so, how can i skip updated this column instead?

thank you.

**Comment by jwoertink** [comment, marked as answer]
If you remove the `licenses_allowed_creation_times` from the `permit_columns`, then it won't set the value from params. 

```crystal
class UpdateUser < User::SaveOperation
  permit_columns password
  
  before_save do
	
	# others password update logic here.
  end
end
```

If you need it in some cases, you could have a separate operation for that.

```crystal
class UpdateUserPassword < User::SaveOperation
  include DefaultUserValidations
  permit_columns password
end

class UpdateUserLicenses < User::SaveOperation
  include DefaultUserValidations
  permit_columns licenses_allowed_creation_times
end
```

You could try something like that.

**Reply by zw963** [reply]
Hi, thanks for your reply.

I thought, when used with avram, we encourage move the saved logic to user save operation, e.g. check if one of column value is empty.

Because, we are use same view template + same kemal action block for update both cases, so,
if follow your guides, if i should write code like this:

```cr
    post path.admin_update_user do |env|
      params = {
        password:                        env.params.body["password"],
        password_confirmation:           env.params.body["password_confirmation"],
      }
	  
	  licenses_allowed_creation_times = env.params.body["licenses_allowed_creation_times"]
	  	  
	  if licenses_allowed_creation_times.present?
	    UpdateUserLicenses.update!(user, **params, licenses_allowed_creation_times: licenses_allowed_creation_times.to_i)
      else
	    UpdateUserPassword.update!(user, **params)
	  end
    end
```

Should we move this `licenses_allowed_creation_times.present?` logic into operations, instead of in current controller?

**Reply by zw963** [reply]
Any update?

**Reply by jwoertink** [reply]
Sorry for the late response. I meant to get back, and forgot. Ok, I think I see what you're trying to do. The logic is better handled within the operation. But separating them is an option.. You may be able to do something like this too:

```crystal
post "/admin" do |env|
  params = {
    password: ...,
    password_confirmation: ...,
   licenses_allowed_creation_times: env.params.body["licenses_allowed_creation_times"]?.presence
  }

  UpdateAdminUser.update!(user, **params)
end

class UpdateAdminUser < User::SaveOperation
  before_save do
    if licenses_allowed_creation_times.value.nil?
    # ignore updating it...
     licenses_allowed_creation_times.value = licenses_allowed_creation_times.original_value
    end
  end
end
```

I'm not sure what the best approach is here, but those are some options for you.

**Reply by zw963** [reply]
Cool, it works!  although, a little hacky.

```cr
licenses_allowed_creation_times.value = licenses_allowed_creation_times.original_value
```

Does this really skip save same value into underline DB? or saved to DB with same value?

That is important, i need know if this hack update the `updated_at`?

**Reply by jwoertink** [reply]
When the record is passed in `user` the attribute sets the `original_value` to what was pulled from the database. Then when a named arg or param value is passed in, that sets the `value`. That way we can see if the value changed like

```crystal
if licenses_allowed_creation_times.changed?(from: "this", to: "that")
  # ...
end
```

by comparing `original_value` and `value`. It should work on `updated_at` too. `updated_at.value = updated_at.original_value` since it won't set the value to that if it already has one

https://github.com/luckyframework/avram/blob/5c553aa7bd67a8ec8f00050eddda8adcdd0a72f8/src/avram/save_operation.cr#L389-L390

**Reply by zw963** [reply]
Hi,  sorry for my bad english.

```
  before_save do
    if licenses_allowed_creation_times.value.nil?
    # ignore updating it...
     licenses_allowed_creation_times.value = licenses_allowed_creation_times.original_value
    end
  end
```

The question is,  when update only one column,  use  ` licenses_allowed_creation_times.value = licenses_allowed_creation_times.original_value`, as above code,  if will caused the `updated_at` column get auto updated to Time.utc?

i want to know, if just save value with original_value like this, if will touch the DB and save same value on DB, and trigger the updated_at hook, and update the value of updated_at ?

**Reply by jwoertink** [reply]
Yes. The `updated_at` will always be updated when `updated_at.value` is nil. A database call is always made even if nothing is being changed.

```crystal
user = UserQuery.first

User::SaveOperation.update(user) do |_, _|
  # still makes a database call and updates updated_at
end
```

**Reply by jwoertink** [reply]
@zw963 I just came up with a new idea on how to clean this up that you can try if you want:

```crystal
UpdateUserLicenses.update!(user, **params, licenses_allowed_creation_times: licenses_allowed_creation_times || Avram::Nothing.new)
```

I wonder if we can just make a global constant for this like:

```
IGNORE_VALUE = Avram::Nothing.new
```

Then if `licenses_allowed_creation_times` is nil, and you want to ignore it instead of using the nil value, you'd just do

```crystal
SomeOp.update(this, some_param: some_param || IGNORE_VALUE)
```

This would obviously cause issues if the value could be a boolean `false`... but something to consider I guess :smile:

**Reply by zw963** [reply]
Okay, i guess if apply the new idea, my code will change to following form: (avram + kemal), and remove the `licenses_allowed_creation_times.value = licenses_allowed_creation_times.original_value` in before_save hook of UpdateUser, right?

```crystal
  post path.admin_update_user do |env|
      user = get_user(env).not_nil!

      licenses_allowed_creation_times = env.params.body["licenses_allowed_creation_times"]

      if licenses_allowed_creation_times.present?
        value = licenses_allowed_creation_times.to_i
      else
        value = Avram::Nothing.new
      end

      params = {
        password:                        env.params.body["password"].as(String),
        password_confirmation:           env.params.body["password_confirmation"].as(String),
        licenses_allowed_creation_times: value,
      }

      UpdateUser.update(user, **params) do |operation, updated_user|
        if operation.saved?
          env.session.destroy
          env.redirect path.admin_users_index
        else
          errors = operation.errors
          render_use_layout "admin/users/edit.ecr"
        end
      end
    end
```

I think it should work, if i am not misunderstood the means of `Avram::Nothing.new`, but we have to move logic to kemal block instead of before_save, i still consider hidden following saving logic into avram, make kemal part is more simple is better?

```
      licenses_allowed_creation_times = env.params.body["licenses_allowed_creation_times"]

      if licenses_allowed_creation_times.present?
        value = licenses_allowed_creation_times.to_i
      else
        value = Avram::Nothing.new
      end
```
